### PR TITLE
feat: add SkipGenFile option

### PR DESCRIPTION
**Pull request by Zxilly**
Close: https://github.com/ent/ent/issues/3508

@rotemtam please review

### CODE REVIEW

File: entproto/extension.go [hunk 1]
```diff
@@ -66,6 +67,16 @@ func WithProtoDir(dir string) ExtensionOption {
 	}
 }
 
+// WithProtoHelper sets whether to generate a helper file next to each .proto file.
+// The helper file contains a //go:generate directive to invoke protoc and compile Go code from the protobuf definitions.
+// If generate.go already exists next to the .proto file, this step is skipped.
+// Set generate to false to disable the generation of the generate.go file.
+func WithProtoHelper(generate bool) ExtensionOption {
+	return func(e *Extension) {
```
Comment (rotemtam) [inline entproto/extension.go:75]: In terms of API I prefer "SkipGenFile" , no args. As we generate the file by default. Also "helper" is too ambiguous, IMO

**Review by rotemtam — CHANGES_REQUESTED** [review]

File: entproto/extension.go (hunk 1 above)
Comment (Zxilly) [reply entproto/extension.go:75]: updated

File: entproto/extension.go [hunk 2]
```diff
@@ -101,11 +112,16 @@ func Hook() gen.Hook {
...
+// Generate takes a *gen.Graph and creates .proto files.
+// Next to each .proto file, Generate creates a generate.go
 // file containing a //go:generate directive to invoke protoc and compile Go code from the protobuf definitions.
 // If generate.go already exists next to the .proto file, this step is skipped.
+// You can use entproto.WithProtoDir to set the directory where the generated .proto files will be written.
+// You can use entproto.SkipGenFile to disable the generation of the generate.go file.
 func Generate(g *gen.Graph) error {
-	x := &Extension{}
+	x := &Extension{
+		skipGenFile: true,
```
Comment (rotemtam) [inline entproto/extension.go:123]: Skip should be false by default (to preserve the current behavior)

**Comment by rotemtam** [comment]
Please see comment , also - please add a unit test verifying this behavior?

Thanks

File: entproto/extension.go (hunk 2 above)
Comment (Zxilly) [reply entproto/extension.go]: sorry, i forget to remove it

**Comment by Zxilly** [comment]
I don't find a previously existing test, where should I add the test file?
Should I add unit tests for individual functions, or should I design an integration test to check if the file is generated?

**Comment by Zxilly** [comment]
Unit test has been added. plz review

File: entproto/extension.go [hunk 3]
```diff
@@ -101,9 +112,12 @@ func Hook() gen.Hook {
 	return x.hook()
 }
 
-// Generate takes a *gen.Graph and creates .proto files. Next to each .proto file, Generate creates a generate.go
+// Generate takes a *gen.Graph and creates .proto files.
+// Next to each .proto file, Generate creates a generate.go
 // file containing a //go:generate directive to invoke protoc and compile Go code from the protobuf definitions.
 // If generate.go already exists next to the .proto file, this step is skipped.
+// You can use entproto.WithProtoDir to set the directory where the generated .proto files will be written.
```
Comment (rotemtam) [inline entproto/extension.go:119]:
```suggestion
```

File: entproto/extension.go [hunk 4]
```diff
@@ -101,9 +112,12 @@ func Hook() gen.Hook {
...
+// You can use entproto.SkipGenFile to disable the generation of the generate.go file.
```
Comment (rotemtam) [inline entproto/extension.go:120]:
```suggestion
// To disable the generation of the generate.go file, use the `entproto.SkipGenFile()` option.
```

File: entproto/internal/todo/generate_test.go [hunk 5]
```diff
@@ -27,9 +27,8 @@ import (
 )
 
 func TestGenerate(t *testing.T) {
-	tgt, err := os.MkdirTemp(os.TempDir(), "entproto-test-*")
-	defer os.RemoveAll(tgt)
-	require.NoError(t, err)
+	tgt := t.TempDir()
+
```
Comment (rotemtam) [inline entproto/internal/todo/generate_test.go:31]:
```suggestion
```

File: entproto/internal/todo/generate_test.go [hunk 6]
```diff
@@ -42,3 +41,24 @@ func TestGenerate(t *testing.T) {
...
+	entprotoExt,err := entproto.NewExtension(entproto.SkipGenFile())
+	require.NoError(t, err)
+
+	err = entc.Generate("./ent/schema",
+		&gen.Config{
+			Target: tgt,
+		},
+		entc.Extensions(entprotoExt))
+	require.NoError(t, err)
```
Comment (rotemtam) [inline entproto/internal/todo/generate_test.go:56]:
```suggestion
	err = entc.Generate(
		"./ent/schema",
		&gen.Config{
			Target: tgt,
		},
		entc.Extensions(entprotoExt),
	)
	require.NoError(t, err)
```

**Review by rotemtam — COMMENTED** [review]
Final style comments and we're good

**Comment by rotemtam** [comment]
hey @Zxilly  CI is red, can you check?

File: entproto/extension.go [hunk 7]
```diff
@@ -66,6 +67,16 @@ func WithProtoDir(dir string) ExtensionOption {
 	}
 }
 
+// SkipGenFile skips generate a helper file next to each .proto file.
+// The helper file contains a //go:generate directive to invoke protoc and compile Go code from the protobuf definitions.
+// If generate.go already exists next to the .proto file, this step is skipped.
+// With this option, you can disable the generation of the generate.go file.
```
Comment (rotemtam) [inline entproto/extension.go:73]:
```suggestion
// SkipGenFile skips the generation a helper file generate.go next to each .proto file.
```

File: entproto/extension.go [hunk 8]
```diff
@@ -138,27 +148,31 @@ func (e *Extension) generate(g *gen.Graph) error {
...
+				}
+				toSchema := filepath.Join(toBase, "schema")
+				contents := protocGenerateGo(fd, toSchema)
+				if err := os.WriteFile(genGoPath, []byte(contents), 0600); err != nil {
+					return fmt.Errorf("entproto: failed generating generate.go file for %q: %w", protoFilePath, err)
+				}
 			}
 		}
 	}
+
```
Comment (rotemtam) [inline entproto/extension.go:175]:
```suggestion

```

**Review by rotemtam — APPROVED** [review]
Thanks for the contribution @Zxilly !

**Pull request merged**